### ISSUE TITLE
Loading/Chunking Large Files fails due to memory limits

### ISSUE THREAD

**Issue by aswinshenoy** [aswinshenoy/BayJDO#14]
![Screenshot_20200713-171633](https://user-images.githubusercontent.com/21276922/87301184-bd193480-c52c-11ea-862d-0c44ea1dc7c1.png)

I tried uploading a mere 396MB video to get this error on my phone.

**Comment by aswinshenoy** [comment]
The problem lies within `getChunksFromFile.js` - 
```js
export default async function getChunksFromFile(file, chunkSize = 6 * 1024 * 1024) {
    if(file)
    {
        return await file.arrayBuffer().then((buffer) => {
            let chunks = [];
            while(buffer.byteLength) {
                chunks.push(buffer.slice(0, chunkSize));
                buffer = buffer.slice(chunkSize, buffer.byteLength);
            }
            return chunks;
        });
    }
    return [];
};
```
The file itself needs to be loaded to the memory twice -> for file.arrayBuffer(), and for chunks = [].

I am thinking whether we can chunk progressively when required, one by one, calling something like `getChunkFromRemaining` and eliminate the need for a storing chunks.

**Comment by MidhunSureshR** [comment]
This issue here is that the receiving peer stores all the blobs in an array and combines them into a big blob when all the chunks have arrived. This essentially means that in order to receive a file, it must first be completely stored in the receiving peer's memory. 
This means the amount of data that can be received is strictly bounded by the available memory. This makes it impractical to send multi-gigabyte files using this method.   

Unfortunately this is a limitation of running web applications within a browser. There is simply no JavaScript API for directly accessing the filesystem. You could try using [Writable Streams](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API/Using_writable_streams) but it is not supported by [Firefox](https://caniuse.com/#feat=mdn-api_writablestream). 
[FileSystem API](https://caniuse.com/#feat=filesystem) would have been suited for such a task but it has been discontinued by [W3C](https://www.w3.org/TR/file-system-api/) (and will probably be removed from future releases of Chrome).

The only straightforward solution would be to install an extension on the users browser. This will grant raw filesystem access and chunks that have been written to drive can be marked for garbage collection. But then again extensions are not supported by most smartphone browsers.

On the positive side of things, there is a proposal for a [Native Filesystem API](https://wicg.github.io/native-file-system/).

**Comment by aswinshenoy** [comment]
Yeah, I had identified this problem earlier. But it's not unsolvable actually (atleast when sufficient storage space is available on devices).

The more serious limitation I believe is actually with the memory (RAM) of the device? We aren't even storing a thing to even local storage at the moment, we are just using memory available to hold the file (the chunks are an JS array, i.e completely on memory) all the while transfer happens...

I read that ArrayBuffer can be stored in IndexedDb, which supports large storage. What we can do is store the chunks in IndexedDb (both for reciever and sender)?  This would mean memory would only be needed during chunking them and recombining them? And with further improvement of the chunking algorithm, this can be further optimized. Such a solution with local caching (via IndexedDb) would also mean now we can add support pause/resume etc. too.

What do you think about it? I guess we can definetly give IndexedDb based caching a try, and I believe it definitely fix few problems with large files..

**Comment by MidhunSureshR** [comment]
To clarify, when I used the term memory, I was in fact talking about RAM. Clearly there are methods of achieving what we want. It's just that none of them work uniformly across all browsers and all devices. Yes we could potentially use IndexedDb to store the chunks and it seems to be the most promising solution so far. But that being said, it still wouldn't scale over to Firefox because of the [2GB group limit](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Browser_storage_limits_and_eviction_criteria). Chrome is more liberal, allowing up to 60% of available space to be used. 

This can be more easily seen by running the following snippet in different browsers:
```js
const foo = async () => {
    if (navigator.storage && navigator.storage.estimate) {
        const quota = await navigator.storage.estimate();
        // quota.usage -> Number of bytes used.
        // quota.quota -> Maximum number of bytes available.
        const percentageUsed = (quota.usage / quota.quota) * 100;
        console.log(`You've used ${percentageUsed}% of the available storage.`);
        const remaining = quota.quota - quota.usage;
        console.log(`You can write up to ${remaining} more bytes.`);
    }
}
foo();
```
I remember mega.nz doing exactly what we want to do. They used a combination of FileSystem API on chrome and an extension on Firefox. Mega did [support Firefox](https://mega.nz/blog_12) later but I have no idea how they did it.

**Comment by aswinshenoy** [comment]
> snippet
I had gone through this [article ](https://web.dev/storage-for-the-web/) and snippet from web.dev earlier. I am pretty aware of the limitations with Firefox, But working on well atleast on 1 browser (the most popular one) is pretty much more than enough v/s failing. Chrome is available for almost all platforms (or atleast other chromium browsers), so that should do.  And almost all production webapps use polyfills / separate strategies for dealing with multiple browsers (mega does it too)

> Mega

Yes, Mega and lot of others achieve it, it just requires some advanced implementation. There is obviously no need for `install an extension on the users browser.` or some other workarounds. Maybe you can also checkout -

- https://github.com/jimmywarting/StreamSaver.js
- https://stackoverflow.com/questions/15994554/download-files-like-mega-co-nz  
- Actual code from Mega - https://eu.static.mega.co.nz/js/download_6.js

**Comment by aswinshenoy** [comment]
> ```js
>         console.log(`You can write up to ${remaining} more bytes.`);
> ```

Also, I have an opinion that we should also show & consider storage space available before allowing/starting file transfer by the user.  A user needs to know the transfer has failed because of lack of space available on his phone (which in no way we can solve),

**Comment by MidhunSureshR** [comment]
StreamSaver seems interesting and looks like it also has a ponyfill for Firefox. Do you think using StreamSaver would be better than using IndexedDb? StreamSaver is emulating a server using service workers. I would assume that all the complexity involved with recombining chunks would be handled by the browser. 

> Also, I have an opinion that we should also show & consider storage space available before allowing/starting file transfer by the user. A user needs to know the transfer has failed because of lack of space available on his phone (which in no way we can 
> solve)

That would certainly be a nice feature to have. But I guess if StreamSaver were being used, this would be implicitly handled by the browser.

**Comment by aswinshenoy** [comment]
> if StreamSaver were being used, this would be implicitly handled by the browser?

Well, still no one is gonna tell them why it failed? and don't you think they should be first told about the storage limitations before letting it fail? A text right at where you select the files saying that - "you have X gb left, you can sent upto Y gb of files" would be clear for the user, isn't it? rather than selecting, and getting it failed. Even if StreamSaver handles the error and we throw a message it failed, its always better from a UX perspective to tell to the user its gonna fail (when we know it) v/s show him once it has failed (and I feel people are gonna be less likely to try again clearing up storage space after a fail v/s telling them in advance)

**Comment by aswinshenoy** [comment]
> Do you think using StreamSaver would be better than using IndexedDb

I have no idea since I haven't tried them. StreamSaver is only a library, and IndexedDb might be used in its underlying implementation, idk!  I perhaps think there is a slight hedge when relying less on a library, IndexedDb is a much more browser thing. I perhaps liked how Mega.nz uses multiple strategies to handle with multiple devices/browser,  we could use IndexedDb, StreamSaver etc. together, and see if something will work for the client.

**Comment by MidhunSureshR** [comment]
>Well, still no one is gonna tell them why it failed?

Wouldn't the browser or OS tell them that there is insufficient disk space? I make no guarantee since I've not run out of disk space in a decade or so :blush:.
>and don't you think they should be first told about the storage limitations before letting it fail?

Of course. It make a lot of sense from a UX perspective. But I just don't think that it can be done if StreamSaver.js were being used. Because unlike IndexedDb, StreamSaver is emulating a normal download by sending Content-Disposition headers to the browser using service workers. This means the burden of saving the file is on the browser. And since browsers have direct disk access, this is of no issue. But to us, as developers writing in JavaScript, the disk does not exist. We cannot know if there is enough space on the device. This is strictly for StreamSaver. 
This [example](https://jimmywarting.github.io/StreamSaver.js/examples/write-slowly.html) from their repo illustrates this much better than I can explain.

On the other hand if IndexedDb were being used, we could definitely know how much space is available.

> StreamSaver is only a library, and IndexedDb might be used in its underlying implementation

I thought so too but that is certainly not what is happening. StreamSaver is a [clever hack](https://github.com/jimmywarting/StreamSaver.js#how-does-it-work).

> I perhaps liked how Mega.nz uses multiple strategies to handle with multiple devices/browser, we could use IndexedDb, StreamSaver etc. together, and see if something will work for the client.

Yeah I like that too. But StreamSaver **does** work across all browsers. Seeing Mega does work on Firefox, there must be some way of overcoming the 2GB limit (or working around it). So if both the solutions can be made to work on all major browsers, is there really a need to switch to different methods depending on the client?

**Comment by aswinshenoy** [comment]
> there is insufficient disk space?

Its actually not the same as user's storage let remind you.  As we had already discussed, and you had pointed out the snippet for calculating the quota space available is infact 60% even in chrome, there is clearly difference! Most devices may throw up storage warnings when the device itself runs out of storage, and not really when browsers quota is finishing?! Infact, when I tried on a device which had no capacity to hold the file its getting, it just fails! no warning (why would even system warn?). We also need to show warning like 'You have no sufficient space in your device to receive it`. Said that, the system won't really warn you when you are just picking the files, because it doesn't expect you are gonna store those chunks in a variable (or in IndexedDb in future), it will fail, but the error will be uncaught (its an error in JS we made, and we need to fix it), and the app simply will crash in production.

> But I just don't think that it can be done if StreamSaver.js were being used.

Sorry, but you missed the point. I am telling you about a simple text that needs to be shown in our UI (once connected) showing how much space is available and thus how much file size is supported for transfer. This is no way related to the actual transfer, and no way related `StreamSaver.js`.

> We cannot know if there is enough space on the device

When during a transfer, we can certainly do that :)  First, before transferring the files we send a normal data (JSON) via webRTC, with the file's meta data (i.e. file size) to the receiver, and expect him to request starting of the transfer. And on the reciever's client, we can actually run the Snippet you pointed out earlier from web.dev, calculate space available on the device (the quota chrome gives us). 

> This means the burden of saving the file is on the browser. And since browsers have direct disk access, this is of no issue.

Sadly, I dont think so, please correct me if I am wrong. Browsers definitely have direct disk access, but there is no way our JS apps can access those (as discussed in the beginning, no web APIs yet) and are limited by what browser allocates us (our quota). Now you might think **what about service workers**  - from my understanding of what service workers are they are just means for us to run scripts in background threads, without interfering with the user interface (a multi-threading approach for performance). Sure they help, but service workers also take JS scripts, and are limited by the same Web APIs that are available for us normally. Service Workers dont give any extra storage permission, correct me if I am wrong.  `developers writing in JavaScript, the disk does not exist` applies to service-worker scripts to since thats afterall just another script written by us (or the people at the library)


> StreamSaver is a [clever hack](https://github.com/jimmywarting/StreamSaver.js#how-does-it-work).

Cant really comment, without trying for our use case. Also, showing transfer rate and progress is a must, even if using a library, so there should be a way for us to do that as well. (flexibility for our needs is also needed)


> So if both the solutions can be made to work on all major browsers, is there really a need to switch to different methods depending on the client?

I didnt understand you completely :)  See we need to write those switch statements based on user agent / device etc. based on our knowledge that solution a works with this agent b and not a. We need to define which solution needs to be used when, obviously, and that requires switching.

**Comment by MidhunSureshR** [comment]
I think you are failing to differentiate between how a browser normally downloads a file versus the various API and mechanisms JavaScript provides to store data on the client side. You can download a file until your hard disk runs out of free space. 
If I had a 1TB HDD and I set the download location to that drive, I can download into that drive until there is no free space left.
None of storage limits and eviction criteria would apply here. StreamSaver does not use up any quota space because it acts like a normal download. 

>Browsers definitely have direct disk access, but there is no way our JS apps can access those

>Sorry, but you missed the point. I am telling you about a simple text that needs to be shown in our UI (once connected) showing how much space is available and thus how much file size is supported for transfer. This is no way related to the actual transfer, and no way related StreamSaver.js.

You're contradicting yourself here. Clearly if our JS apps cannot access the disk then it follows that there is no way of determining the amount of free space. StreamSaver is not using any webstorage (i.e IndexedDb, Cookies, Cache and the like). If you really wanted to show such a message you would have to use IndexedDb instead of StreamSaver. 

>I didnt understand you completely :) See we need to write those switch statements based on user agent / device etc. based on our knowledge that solution a works with this agent b and not a. We need to define which solution needs to be used when, obviously, and that requires switching.

When Mega was launched there was no one solution that worked across all browsers. Therefore they used different solutions on different browsers. But that is not the case here. StreamSaver does work on all browsers. Therefore there really isn't a need to use different methods on different browsers.

**Comment by aswinshenoy** [comment]
> You can download a file until your hard disk runs out of free space.

Dear brother, you are not just downloading a file. PLEASE note that this issue itself starts when you are chunking a large file for sending to the receiver.  You are not just saving a file, you also need to serve it, and while you are serving the file - you also need to hold the file, the chunks of the file in memory / storage while it is being sent from the sender. Currently, its in the memory as JS variable and hence is failing, the solution we need is to temporarily hold it in some sort of storage. But even for that storage, you have quota thats allocated by Chrome, and it would be good if say to them through a text that you have x space left/allowed before they select a file, so that a failure due to this exhaustion of quota. I clearly, multiple times, told that this indication is for the person sending, and not really for the reciever.

> Clearly if our JS apps cannot access the disk then it follows that there is no way of determining the amount of free space.

Sir, JS apps cannot access disk, but from the snippet you yourself posted earlier (from web.dev), JS apps always have a method to know how much space (the quota, and not the actual storage) they can utilize to write (not even access), and we are interested only in that. Please don't confuse yourself here between access to disk vs/ knowing quota size, I am clearly talking about knowing the total disk quota that chrome allows (like 60% of drive) for the app and not the total storage space available, let alone access to data. 

> When Mega was launched there was no one solution that worked across all browsers.

Cannot comment on Mega. But, if you can achieve to solve the issue of receiving just with StreamSaver across all platform, meeting all the flexibility we need, yes, that should **do**. I am still unclear of how we would be sending the file, before even receiving, because once chunking is done, you need to hold that somewhere (and not clearly memory),  this end of the problem requires a medium to store, wherein I pointed out IndexedDb. This is what this issue is about, the primary issue is about how it fails when trying to send because the chunks cant be store.

**Comment by MidhunSureshR** [comment]
No hard feelings dude. :smile: :v:

>I am clearly talking about knowing the total disk quota that chrome allows (like 60% of drive) for the app and not the total storage space available

You mean calculating the free space from the allotted space mathematically? Like this?
![equation](https://latex.codecogs.com/png.latex?%5Cfn_jvn%20%5Csmall%20totalSpace%20%3D%20allotedSpace%20*%20%5Cfrac%7B100%7D%7B60%7D)
If that's what you meant, it's good thinking but it still wouldn't work because WebStorage location and Download location can be different. More precisely WebStorage is always in your profile directory. For example, in Firefox, that would be:
`
%AppData%\Roaming\Mozilla\Firefox\Profiles\your profile id\indexedDB\domain
`
While Downloads could be stored at:
`
D:\Downloads
`
That is the quota reveals no information about total space that can be used for downloads. 

But none of this is necessary when StreamSaver is used. 

When I use bayjdo.com and I receive a file there are two things happening:
(A): The file is received and stored as an array of blobs in receiving peer (and later combined to form a big blob of data).
I see a progress when this is happening. 
(B): After (A) has completed, I click on the Save button to save the file to my disk.

So what is happening when (B) occurs? URL.createObjectURL is used to create the DOMString and it is assigned to the href attribute of a dynamically created anchor tag. Once the user clicks on the button, the browser handles saving the file to disk.
The browser does this using streams (not the JavaScript streams, the good ol' C++ low-level streams) so that chunks can be progressively written to file and then immediately released from memory. 

What StreamSaver does is it tricks the browser to do (B). This means that we do not need to do (A). We don't need to hold the chunks in memory, we can give it to the browser and remove all references to it so that the garbage collector can release this memory.

The memory overflow seems to happen on the receiver side and not on the senders side.

**Comment by aswinshenoy** [comment]
> Download

I am NOT talking about DOWNLOAD. I am not telling about the need to show the space availability for the receiver, but instead for the person who is going to send the file. Please try to understand.

**Comment by aswinshenoy** [comment]
> When I use bayjdo.com and I receive a file there are two things happening:

Someone or some device is sending you, I am talking about that, again and again. Forget about saving and receiving for a moment!
Okay, so for the file event to start transmitting, the sender's device should chunk the files right? and after chunking, the sender's phone should temporarily hold it right? now where is it holding? currently in a JS var, thus cause the memory issue (RAM is exhausting), so that is exactly where and why I want to introduce a storage medium to hold the data until the transfer is complete. Hope you understand.

**Comment by aswinshenoy** [comment]
> The memory overflow seems to happen on the receiver side and not on the senders side.

Can you tell me how it is not (or wont be) happening on the sender's side please?

**Comment by aswinshenoy** [comment]
> What StreamSaver does is it tricks the browser to do (B). This means that we do not need to do (A). We don't need to hold the chunks in memory, we can give it to the browser and remove all references to it so that the garbage collector can release this memory.

I haven't got to learn much about it let alone try, if you can use it solve the "saving & receiving problem" we have, great. I maybe unable to comprehend how it works underlying, but since you are convinced, its definitely worth a try and would then certainly solve one end of the problem.

Hope you shall also soon recognize the other end of the problem as well 👍

**Comment by MidhunSureshR** [comment]
>Can you tell me how it is not (or wont be) happening on the sender's side please? On both sides, chunks are being held, on the receivers end perhaps, its getting progressively build up, while on the sender's side the whole set of chunks (we made from the file) is being held in a JS var.

There is no memory overflow on the senders side. This can be verified by running a stress test and measuring memory usage via dev tools (you can also use a heap snapshot but this will suffice for now). 
![Screenshot_of_test](https://i.imgur.com/NVCIOew.png)

The sender is using about 100MB of memory while the receiver is using more than 7GB.
This is because the File object (special kind of blob) returned from the Open File dialog (using [input type="file" ](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file) ) is not in memory but rather stored in disk. 
You use Dropzone.js which uses the following in its underlying implementation:
 ```html
 <input type="file">
``` 
Here is a snippet from your chunking mechanism:
```js
  _getNextSlice = async () => {
        const slice = this._file.slice(
            this._offset,
            this._offset + this._chunkSize
        );
        return await slice.arrayBuffer().then((buffer) => buffer);
    };
```
The chunk is brought into memory only when `slice.arrayBuffer()` is called. Once the chunk is transmitted to the receiving peer, no further references to it exist. This causes the GC to release the memory previously occupied by the chunk.

**Comment by aswinshenoy** [comment]
> ![Screenshot_20200713-171633](https://user-images.githubusercontent.com/21276922/87301184-bd193480-c52c-11ea-862d-0c44ea1dc7c1.png)
> 
> I tried uploading a mere 396MB video to get this error on my phone.

Can you help me in understanding this too then? This issue is solved when this solved, I got this on the phone I was sending from, I have no idea why if you say there is no client side issue :)

**Comment by aswinshenoy** [comment]
> There is no memory overflow on the senders side.

Can you explain to me this message - "array buffer allocation failed" and its cause that I got no sending the file on the sender's client? I have shared the screenshot above. However, knowing this I updated the chunking script which may have solved the issue for now, but I dont believe the current script is a fool-proof solution yet, otherwise this issue itself can be closed. 

Before
```

export default async function getChunksFromFile(file, chunkSize = 6 * 1024 * 1024) {
    if(file) {
        return await file.arrayBuffer().then((buffer) => {
            let chunks = [];
            while(buffer.byteLength) {
                chunks.push(buffer.slice(0, chunkSize));
                buffer = buffer.slice(chunkSize, buffer.byteLength);
            }
            return chunks;
        });
    }
    return [];
};
```

After
```
  _getNextSlice = async () => {
        const slice = this._file.slice(
            this._offset,
            this._offset + this._chunkSize
        );
        return await slice.arrayBuffer().then((buffer) => buffer);
    };
```
If you think the new algorithm I had earlier introduced solves this issue already (since thats evident on your stress test), we can pretty much close this issue - this issue is particularly loading/chunking large files failing with `array buffer allocation failed`.

**Comment by aswinshenoy** [comment]
> Receiver is using more than 7GB ( which will increase as more chunks accumulate).

Yes, I completely acknowledge that the issue obviously exists for the reciever too, but that is separate issue, and we can try to solve that using StreamSaver.js like you told earlier. If the sender is passing all stress tests already, well, this issue - #14 seems to be already solved :)

**Comment by aswinshenoy** [comment]
> The chunk is brought into memory only when `slice.arrayBuffer()` is called. Once the chunk is transmitted to the receiving peer, no further references to it exist. This causes the GC to release the memory previously occupied by the chunk.

Yes, totally my bad, I had forgotten that I had already patched it with a new algorithm (from fileChunker.js) and made the `getChunksFromFile.js` obsolete.

**Comment by MidhunSureshR** [comment]
Yep. This issue can be closed.

**Comment by aswinshenoy** [comment]
Ok, closing for now since screenshots of stress tests submitted by @MidhunSureshR  shows chunking of extremely large file like of 26GB does not seem to fail now, though receiving end may have issues in holding it. However, #18 needs a solution to solve the larger issue #10.
